TESTING.md

```diff
@@ -0,0 +1,69 @@
+<!-- TOC -->
+- [Testing](#testing)
+  - [Code Coverage Collecting](#code-coverage-collecting)
+    - [OpenSearch plugin](#opensearch-plugin)
+      - [Gradle project](#gradle-project)
+      - [Maven project](#maven-project)
+      - [Limitation](#limitation)
+    - [OpenSearch Dashboards plugin](#opensearch-dashboards-plugin)
+  - [Code Coverage Reporting](#code-coverage-reporting)
+    - [Code coverage report upload through CI workflow](#code-coverage-report-upload-through-ci-workflow)
+    - [A status badge in README file to show the code coverage](#a-status-badge-in-readme-file-to-show-the-code-coverage)
+    - [Codecov integration in Pull Requests](#codecov-integration-in-pull-requests)
+<!-- TOC -->
+
+## Testing
+
+### Code Coverage Collecting
+Code coverage is a measurement of how many lines of code are executed under the test suites.
+
+#### OpenSearch plugin
+It is recommended to use [JaCoCo](https://www.eclemma.org/jacoco/) library to generate the code coverage report. The following guide may help you use JaCoCo in OpenSearch plugins with ease.
+##### Gradle project
+1. Add [JaCoCo Gradle plugin](https://docs.gradle.org/current/userguide/jacoco_plugin.html) to `build.gradle` in your repository. [example](https://github.com/opensearch-project/sql/blob/opensearch-1.0.0.0-rc1/build.gradle#L48)
+2. Set Jacoco to export coverage report in `xml` format. [example](https://github.com/opensearch-project/sql/blob/opensearch-1.0.0.0-rc1/build.gradle#L101-L105)
+3. Call gradle task "jacocoTestReport" to generate report (`./gradlew jacocoTestReport`) after running your tests. (Or add a `dependsOn` to make "jacocoTestReport" task depending on your test tasks. [example](https://github.com/opensearch-project/index-management/blob/opensearch-1.0.0.0-beta1/build-tools/coverage.gradle))
+
+##### Maven project
+Since OpenSearch uses Gradle as the build tool, we don't recommend using Maven.
+Code coverage for a Maven project can be generated by [JaCoCo Maven plugin](https://www.eclemma.org/jacoco/trunk/doc/maven.html), and here is an [example](https://github.com/opensearch-project/security/blob/v1.0.0.0-rc1/pom.xml#L636).
+
+##### Limitation
+1. JaCoCo Gradle plugin can not collect code coverage properly for integration tests (tests that extend `OpenSearchIntegTestCase` or `OpenSearchRestTestCase`, see [here](https://github.com/opensearch-project/OpenSearch/blob/1.0/TESTING.md#base-classes-for-test-cases) to learn more). However, there is a workaround to collect coverage when the integration test is running on a local single node cluster. For your reference, [this file](https://github.com/opensearch-project/index-management/blob/opensearch-1.0.0.0-beta1/build-tools/coverage.gradle) is a custom Gradle pluign to help achieve, and it can be applied conditionally like [here](https://github.com/opensearch-project/index-management/blob/opensearch-1.0.0.0-beta1/build.gradle#L90-L93).
```

Comment:
Super nit: 
`pluign` to `plugin`
`gralde` to `gradle`

Reply:
Oh, thank you! 😳 You are so careful!

Reply:
Each kind of mistake spelling has got 2 samples, so terrible. 🤡